src/quest/historian.py

```diff
@@ -877,16 +877,19 @@ async def _run(self, *args, **kwargs):
             self._open_tasks.append(external_task)
             external_task.add_done_callback(lambda t: self._open_tasks.remove(t) if t in self._open_tasks else None)
 
-            return await self.start_task(
+            task = await self.start_task(
                 self._run_with_exception_handling,
                 *args, **kwargs,
                 name=f'{self.workflow_id}.main',
                 task_factory=_task_factory.create_task
             )
+            self._history.clear()
```

Comment:
I think the `self._history.clear()` should go in `_run_with_exception_handling`.

Also, please move the `self._workflow_completed = True` and `await self._handle_resource...` from `_run_with_args` to `run_with_exception_handling`.